File: cond-catch.py
#!/usr/bin/env python3

# Conditional Match (or Re-Raise)

from time import thread_time_ns


class EmptyException(Exception):
    pass


should_catch = False


def main(argv):
    times = 1
    if 1 < len(argv):
        times = int(argv[1])
    if 2 < len(argv):
        should_catch = 0 < int(argv[2])

    start_time = thread_time_ns()
    for count in range(times):
        try:
            try:
                raise EmptyException()
            except EmptyException as exc:
                if not should_catch:
                    raise
        except EmptyException:
            pass

    end_time = thread_time_ns()
    print('Run-Time (s): {:.1f}'.format((end_time - start_time) / 1_000_000_000.))


if '__main__' == __name__:
    import sys
    main(sys.argv)

File: CondCatch.java
// Conditional Match (or Re-Raise)

class EmptyException extends Exception {}

public class CondCatch {
	static boolean should_catch = false;

	private static long loop(int times) {
		long startTime = System.nanoTime();
		for (int count = 0 ; count < times ; ++count) {
			try {
				try {
					throw new EmptyException();
				} catch (EmptyException exc) {
					if (!should_catch) {
						throw exc;
					}
				}
			} catch (EmptyException exc) {
				// ...
			}
		}
		long endTime = System.nanoTime();
		return endTime - startTime;
	}

	public static void main(String[] args) {
		int times = 1;
		if (0 < args.length) {
			times = Integer.parseInt(args[0]);
		}
		if (1 < args.length) {
			should_catch = 0 != Integer.parseInt(args[1]);
		}

		// Warm-Up:
		loop(1000);

		long time = loop(times);
		System.out.format("Run-Time (s): %.1f%n", time / 1_000_000_000.);
	}
}
